src/components/ActionSheetAwareScrollView/index.ios.js

```diff
@@ -0,0 +1,527 @@
+ * It sends KEYBOARD_CLOSE action. But we ignore it since our emojiPickerOpen state can only handle
+ * EMOJI_PICKER_CLOSE action. so we write the logic for handling hiding the keyboard
+ * but maintaining the offset based on the keyboard state shared value
+ * 7. we close the picker and send EMOJI_PICKER_CLOSE action which transitions us back into keyboardOpen state.
+ *
+ * State machine object example:
+ * const stateMachine = {
+ *   idle: {
+ *       KEYBOARD_OPEN: 'keyboardOpen',
+ *   },
+ *   keyboardOpen: {
+ *       KEYBOARD_CLOSE: 'idle',
+ *       EMOJI_PICKER_OPEN: 'emojiPickerOpen',
+ *   },
+ *   emojiPickerOpen: {
+ *       EMOJI_PICKER_CLOSE: 'keyboardOpen',
+ *   },
+ * }
+ *
+ * Initial state example:
+ * {
+ *     previous: null,
+ *     current: {
+ *         state: 'idle',
+ *         payload: null,
+ *     },
+ * }
+ *
+ * @param {Object} stateMachine - a state machine object
+ * @param {Object} initialState - the initial state of the state machine
+ * @returns {Object} - an object containing the current state, a transition function, and a reset function
+ */
+function useStateMachine(stateMachine, initialState) {
+    const currentState = useSharedValue(initialState);
+
+    const transitionWorklet = useWorkletCallback((action) => {
+        if (!action) {
+            throw new Error('state machine action is required');
+        }
+
+        const state = currentState.value;
+
+        log(`Current STATE: ${state.current.state}`);
+        log(`Next ACTION: ${action.type}`, action.payload);
+
+        const nextMachine = stateMachine[state.current.state];
+
+        if (!nextMachine) {
+            log(`No next machine found for state: ${state.current.state}`);
+            return;
+        }
+
+        const nextState = nextMachine[action.type];
+
+        if (!nextState) {
+            log(`No next state found for action: ${action.type}`);
+            return;
+        }
+
+        // eslint-disable-next-line no-nested-ternary
+        const nextPayload = typeof action.payload === 'undefined' ? state.current.payload : nextState === state.current.state ? {
+            ...state.current.payload,
+            ...action.payload,
+        } : action.payload;
+
+        log(`Next STATE: ${nextState}`, nextPayload);
+
+        currentState.value = {
+            previous: state.current,
+            current: {
+                state: nextState,
+                payload: nextPayload,
+            },
+        };
+    });
+
+    const resetWorklet = useWorkletCallback(() => {
+        log('RESET STATE MACHINE');
+        currentState.value = initialState;
+    }, []);
+
+    const reset = useCallback(() => {
+        runOnUI(resetWorklet)();
+    }, []);
+
+    const transition = useCallback((action) => {
+        runOnUI(transitionWorklet)(action);
+    }, []);
+
+    return {
+        currentState,
+        transitionWorklet,
+        transition,
+        reset,
+        resetWorklet,
```

Comment:
Looks like `resetWorklet` is not used anywhere outside this function.

Reply:
It's now own separated hook so keeping this method as "exported" for it makes more sense.